### PR TITLE
refactor api compatibility container creation to specgen

### PR DESCRIPTION
**Pull request by baude**
when using the compatibility layer to create containers, it used code paths to the pkg/spec which is the old implementation of containers.  it is error prone and no longer being maintained.  rather that fixing things in spec, migrating to specgen usage seems to make the most sense.  furthermore, any fixes to the compat create will not need to be ported later.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: cmd/podman/common/create_opts.go [hunk 1]
```diff
@@ -111,3 +121,276 @@ type ContainerCLIOpts struct {
...
+			hostport, err := strconv.Atoi(pb.HostPort)
+			if err != nil {
+				return nil, nil, err
+			}
+			tmpPort := specgen.PortMapping{
+				HostIP:        pb.HostIP,
+				ContainerPort: uint16(port.Int()),
+				HostPort:      uint16(hostport),
+				Range:         0,
+				Protocol:      "", // TODO we have no protocol information on ports here. Should we do both?
```
Comment (mheon) [inline cmd/podman/common/create_opts.go:230]: I think protocol might be container in ContainerPort?

Comment (mheon) [reply cmd/podman/common/create_opts.go:230]: Contained in*

**Comment by mheon** [comment]
One thing, otherwise LGTM. Less `pkg/spec` makes me very happy.

**Comment by baude** [comment]
fixed

File: cmd/podman/common/create_opts.go [hunk 2]
```diff
@@ -111,3 +121,276 @@ type ContainerCLIOpts struct {
...
+		Rm:                cc.HostConfig.AutoRemove,
+		SecurityOpt:       cc.HostConfig.SecurityOpt,
+		ShmSize:           strconv.Itoa(int(cc.HostConfig.ShmSize)),
+		StopSignal:        cc.Config.StopSignal,
+		StoreageOpt:       stringMaptoArray(cc.HostConfig.StorageOpt),
+		Sysctl:            stringMaptoArray(cc.HostConfig.Sysctls),
+		Systemd:           "true", // podman default
+		TmpFS:             stringMaptoArray(cc.HostConfig.Tmpfs),
+		TTY:               cc.Config.Tty,
+		//Ulimit:            cc.HostConfig.Ulimits,            // ask dan, no documented format
```
Comment (rhatdan) [inline cmd/podman/common/create_opts.go:351]: What happens when you uncomment this?

**Comment by rhatdan** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM
assuming happy tests

**Comment by mheon** [comment]
One APIv2 test is failing

File: test/apiv2/20-containers.at [hunk 3]
```diff
@@ -207,7 +207,7 @@ t DELETE containers/$cid 204
 t DELETE containers/$cid_top 204
 
 # test the apiv2 create, shouldn't ignore the ENV and WORKDIR from the image
-t POST containers/create '"Image":"'$ENV_WORKDIR_IMG'","Env":["testKey1"]' 201 \
+t POST containers/create '"Image":"'$ENV_WORKDIR_IMG'","Env":["testKey1=foo"]' 201 \
```
Comment (zhangguanzhang) [inline test/apiv2/20-containers.at:210]: https://github.com/containers/podman/blob/93af493ccb13b3d93717bf9c805f820c9e3a2fba/test/apiv2/test-apiv2#L154-L163
the line `sed -e 's/\(.*\)=\(.*\)/"\1":"\2"/' <<<$i` will convert `=` to `:`, so don't need `testKey1=foo`

File: cmd/podman/common/create_opts.go (hunk 2 above)
Comment (nalind) [reply cmd/podman/common/create_opts.go:357]: Most likely this corresponds to the argument values described in https://github.com/containers/buildah/blob/master/docs/buildah-bud.md.

File: test/apiv2/20-containers.at [hunk 4]
```diff
@@ -206,16 +206,6 @@ t POST containers/${cid_top}/stop "" 204
...
-# test the apiv2 create, shouldn't ignore the ENV and WORKDIR from the image
-t POST containers/create '"Image":"'$ENV_WORKDIR_IMG'","Env":["testKey1"]' 201 \
-  .Id~[0-9a-f]\\{64\\}
-cid=$(jq -r '.Id' <<<"$output")
-t GET containers/$cid/json 200 \
-  .Config.Env~.*REDIS_VERSION= \
-  .Config.Env~.*testKey1= \
-  .Config.WorkingDir="/data" # default is /data
-t DELETE containers/$cid 204
-
```
Comment (zhangguanzhang) [inline test/apiv2/20-containers.at:218]: cannot delete these lines

Comment (edsantiago) [reply test/apiv2/20-containers.at:218]: @zhangguanzhang the problem is that when you submitted your initial PR, APIv2 tests were disabled in CI. They weren't running at all. When we tried to reenable those tests, the newly-added ones broke. I never reviewed that PR, and right now it is hard to guess the original intention, so for now the best approach is just to delete them and then add something back later that will work. I am really sorry for the hassle.

Comment (zhangguanzhang) [reply test/apiv2/20-containers.at:218]: ok, I will submit these after enabled APIv2 test

**Comment by zhangguanzhang** [comment]
LGTM

**Pull request merged**